File: typing.js
// typing.js
#!/usr/bin/env node
'use strict';

const fs = require('fs');
const path = require('path');
const os = require('os');
const readline = require('readline');

const COLORS = {
    reset: '\x1b[0m',
    green: '\x1b[92m',
    red: '\x1b[91m',
    yellow: '\x1b[93m',
    blue: '\x1b[94m',
    bold: '\x1b[1m'
};

function colorize(text, color) {
    return COLORS[color] + text + COLORS.reset;
}

const TEXTS = {
    easy: [
        'кот и собака играют во дворе',
        'солнце светит ярко сегодня',
        'быстрый лис прыгает через забор'
    ],
    medium: [
        'программирование это искусство создавать алгоритмы для решения задач',
        'в мире информационных технологий важно постоянно учиться новому',
        'компьютерные науки охватывают множество направлений'
    ],
    hard: [
        'разработка программного обеспечения требует глубоких знаний и навыков работы в команде',
        'современные технологии изменяют мир и открывают новые возможности для человечества',
        'искусственный интеллект и машинное обучение становятся основой будущих инноваций'
    ]
};

class TypingTest {
    constructor(level = 'medium') {
        this.level = level;
        const list = TEXTS[level];
        this.text = list[Math.floor(Math.random() * list.length)];
        this.userInput = '';
        this.startTime = null;
        this.endTime = null;
        this.statsFile = path.join(os.homedir(), '.typing_stats.json');
        this.loadStats();
    }

    loadStats() {
        try {
            this.stats = JSON.parse(fs.readFileSync(this.statsFile, 'utf8'));
        } catch {
            this.stats = {};
        }
    }

    saveStats() {
        fs.writeFileSync(this.statsFile, JSON.stringify(this.stats, null, 2));
    }

    displayText() {
        console.log(colorize('\nНапечатайте следующий текст:', 'bold'));
        console.log(colorize(this.text, 'blue'));
        console.log('\nНажмите Enter после ввода. Время начнётся с первой буквы.');
        return new Promise(resolve => {
            const rl = readline.createInterface({
                input: process.stdin,
                output: process.stdout
            });
            rl.question('Готовы? Нажмите Enter...', () => {
                rl.close();
                resolve();
            });
        });
    }

    async getUserInput() {
        console.log('\nНачинайте печатать:');
        const rl = readline.createInterface({
            input: process.stdin,
            output: process.stdout
        });
        this.startTime = Date.now();
        const input = await new Promise(resolve => {
            rl.question('', answer => {
                resolve(answer);
                rl.close();
            });
        });
        this.userInput = input;
        this.endTime = Date.now();
    }

    calculateAndDisplay() {
        const original = this.text;
        const typed = this.userInput;
        const total = original.length;
        const typedLen = typed.length;
        let correct = 0;
        let errors = 0;
        for (let i = 0; i < total; i++) {
            if (i < typedLen && original[i] === typed[i]) {
                correct++;
            } else {
                errors++;
            }
        }
        if (typedLen < total) errors += total - typedLen;
        const elapsed = (this.endTime - this.startTime) / 1000;
        const minutes = elapsed / 60;
        const wpm = (correct / 5) / minutes;
        const accuracy = (correct / total) * 100;

        console.log(colorize('\n📊 Результаты:', 'bold'));
        console.log(`  Скорость: ${wpm.toFixed(1)} WPM`);
        console.log(`  Точность: ${accuracy.toFixed(1)}%`);
        console.log(`  Правильных символов: ${correct}/${total}`);
        console.log(`  Ошибок: ${errors}`);
        console.log(`  Время: ${elapsed.toFixed(2)} сек`);

        if (!this.stats[this.level] || wpm > this.stats[this.level]) {
            this.stats[this.level] = wpm;
            this.saveStats();
            console.log(colorize('🏆 Новый рекорд для этого уровня!', 'green'));
        } else {
            console.log(`Лучший результат: ${this.stats[this.level].toFixed(1)} WPM`);
        }
    }

    async run() {
        console.log(colorize('⌨️  Тест скорости печати', 'bold'));
        console.log(`Уровень: ${this.level}`);
        await this.displayText();
        await this.getUserInput();
        this.calculateAndDisplay();
    }
}

async function main() {
    let level = 'medium';
    let showStats = false;
    let resetStats = false;
    let customText = '';
    const args = process.argv.slice(2);
    for (let i = 0; i < args.length; i++) {
        const arg = args[i];
        if (arg === 'easy' || arg === 'medium' || arg === 'hard') {
            level = arg;
        } else if (arg === '-t' || arg === '--text') {
            if (i+1 < args.length) {
                customText = args[++i];
            }
        } else if (arg === '-s' || arg === '--stats') {
            showStats = true;
        } else if (arg === '-r' || arg === '--reset') {
            resetStats = true;
        } else if (arg === '-h' || arg === '--help') {
            console.log('Usage: node typing.js [easy|medium|hard] [-t text] [-s] [-r]');
            return;
        }
    }
    if (resetStats) {
        const f = path.join(os.homedir(), '.typing_stats.json');
        if (fs.existsSync(f)) fs.unlinkSync(f);
        console.log('Статистика сброшена.');
        return;
    }
    if (showStats) {
        const f = path.join(os.homedir(), '.typing_stats.json');
        try {
            const stats = JSON.parse(fs.readFileSync(f, 'utf8'));
            console.log(colorize('📊 Статистика:', 'bold'));
            for (const [lvl, wpm] of Object.entries(stats)) {
                console.log(`  ${lvl}: ${wpm.toFixed(1)} WPM`);
            }
        } catch {
            console.log('Статистика пуста.');
        }
        return;
    }
    const test = new TypingTest(level);
    if (customText) {
        test.text = customText;
    }
    await test.run();
}

main().catch(console.error);

File: typing.py
# typing.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import sys
import os
import random
import json
import time
from pathlib import Path

# ANSI-цвета
COLORS = {
    'reset': '\033[0m',
    'green': '\033[92m',
    'red': '\033[91m',
    'yellow': '\033[93m',
    'blue': '\033[94m',
    'bold': '\033[1m'
}

def colorize(text, color):
    return f"{COLORS.get(color, '')}{text}{COLORS['reset']}"

# Встроенные тексты по уровням
TEXTS = {
    'easy': [
        "кот и собака играют во дворе",
        "солнце светит ярко сегодня",
        "быстрый лис прыгает через забор"
    ],
    'medium': [
        "программирование это искусство создавать алгоритмы для решения задач",
        "в мире информационных технологий важно постоянно учиться новому",
        "компьютерные науки охватывают множество направлений"
    ],
    'hard': [
        "разработка программного обеспечения требует глубоких знаний и навыков работы в команде",
        "современные технологии изменяют мир и открывают новые возможности для человечества",
        "искусственный интеллект и машинное обучение становятся основой будущих инноваций"
    ]
}

class TypingTest:
    def __init__(self, level='medium'):
        self.level = level
        self.text = random.choice(TEXTS[level])
        self.start_time = None
        self.end_time = None
        self.user_input = ""
        self.stats_file = Path.home() / '.typing_stats.json'
        self.load_stats()

    def load_stats(self):
        if self.stats_file.exists():
            with open(self.stats_file, 'r') as f:
                self.stats = json.load(f)
        else:
            self.stats = {}

    def save_stats(self):
        with open(self.stats_file, 'w') as f:
            json.dump(self.stats, f, indent=2)

    def display_text(self):
        print(colorize("\nНапечатайте следующий текст:", 'bold'))
        print(colorize(self.text, 'blue'))
        print("\nНажмите Enter после ввода. Время начнётся с первой буквы.")
        input("Готовы? Нажмите Enter...")
        print("\nНачинайте печатать:")

    def get_user_input(self):
        self.start_time = time.time()
        self.user_input = input()
        self.end_time = time.time()

    def calculate_results(self):
        original = self.text
        typed = self.user_input
        total_chars = len(original)
        typed_chars = len(typed)
        correct = 0
        errors = 0
        for i, ch in enumerate(original):
            if i < typed_chars and ch == typed[i]:
                correct += 1
            else:
                errors += 1
        # Если текст короче оригинала, считаем остальные как ошибки
        if typed_chars < total_chars:
            errors += total_chars - typed_chars
        elapsed = self.end_time - self.start_time
        minutes = elapsed / 60.0
        # Количество слов: делим на 5 (стандарт)
        word_count = correct / 5.0
        wpm = word_count / minutes if minutes > 0 else 0
        accuracy = (correct / total_chars) * 100 if total_chars > 0 else 0
        return {
            'wpm': wpm,
            'accuracy': accuracy,
            'correct': correct,
            'errors': errors,
            'total': total_chars,
            'time': elapsed
        }

    def display_results(self, results):
        print(colorize("\n📊 Результаты:", 'bold'))
        print(f"  Скорость: {results['wpm']:.1f} WPM")
        print(f"  Точность: {results['accuracy']:.1f}%")
        print(f"  Правильных символов: {results['correct']}/{results['total']}")
        print(f"  Ошибок: {results['errors']}")
        print(f"  Время: {results['time']:.2f} сек")

        # Сохранение лучшего результата
        if self.level not in self.stats or results['wpm'] > self.stats[self.level]:
            self.stats[self.level] = results['wpm']
            self.save_stats()
            print(colorize("🏆 Новый рекорд для этого уровня!", 'green'))
        else:
            print(f"Лучший результат: {self.stats.get(self.level, 0):.1f} WPM")

    def run(self):
        print(colorize("⌨️  Тест скорости печати", 'bold'))
        print(f"Уровень: {self.level}")
        self.display_text()
        self.get_user_input()
        results = self.calculate_results()
        self.display_results(results)

def main():
    level = 'medium'
    show_stats = False
    reset_stats = False
    custom_text = None
    args = sys.argv[1:]
    for i, arg in enumerate(args):
        if arg in ['easy', 'medium', 'hard']:
            level = arg
        elif arg == '-t' or arg == '--text':
            if i+1 < len(args):
                custom_text = args[i+1]
        elif arg == '-s' or arg == '--stats':
            show_stats = True
        elif arg == '-r' or arg == '--reset':
            reset_stats = True
        elif arg == '-h' or arg == '--help':
            print("Usage: typing.py [easy|medium|hard] [-t text] [-s] [-r]")
            return
    if reset_stats:
        stats_file = Path.home() / '.typing_stats.json'
        if stats_file.exists():
            stats_file.unlink()
        print("Статистика сброшена.")
        return
    if show_stats:
        stats_file = Path.home() / '.typing_stats.json'
        if stats_file.exists():
            with open(stats_file, 'r') as f:
                stats = json.load(f)
                print(colorize("📊 Статистика:", 'bold'))
                for lvl, wpm in stats.items():
                    print(f"  {lvl}: {wpm:.1f} WPM")
        else:
            print("Статистика пуста.")
        return
    if custom_text:
        # Используем пользовательский текст
        test = TypingTest(level)
        test.text = custom_text
        test.run()
    else:
        test = TypingTest(level)
        test.run()

if __name__ == '__main__':
    try:
        main()
    except KeyboardInterrupt:
        print(colorize("\nТест прерван.", 'yellow'))
        sys.exit(0)
